tests/QuaternionTest.mpl

```diff
@@ -0,0 +1,175 @@
+
+"Quaternion" use
+
+QuaternionTest: [];
+
+equal: [
+  v0: v1:;;
+  epsilon: 1.0e-6 v0 cast;
+  v0 v1 - abs epsilon <
+];
+
+equal: [
+  v0: v1:;;
+  v0 vector?
+  v1 vector? and
+  v0 fieldCount v1 fieldCount = and
+] [
+  v0: v1:;;
+  TRUE
+  v0 fieldCount [
+    i v0 @ i v1 @ equal and
+  ] times
+] pfunc;
+
+equal: [
+  q0: q1:;;
+  q0 "QUATERNION" has
+  q1 "QUATERNION" has and
+] [
+  q0: q1:;;
+  q2: q1.entries neg quaternion;
+  q0.entries q1.entries equal q0.entries q2.entries equal or
+] pfunc;
+
+checkEqual: [
+  value0: value1:;;
+  [value0 value1 equal] "value0 and value1 differ" ensure
+];
+
+# basic ops
+[
+  q0: (1.0 2.0 neg 3.0 4.0    ) quaternion;
+  q1: (5.0 6.0     7.0 8.0 neg) quaternion;
+
+  q: (0.182574185835055 0.365148371670111 neg 0.5477225575051 0.7302967433402) quaternion;
+  tests: (
+    {op: [q0 q1 +  ]; expected: ( 6.0  4.0     10.0      4.0 neg) quaternion;}
+    {op: [q0 2.0 * ]; expected: ( 2.0  4.0 neg  6.0      8.0    ) quaternion;}
+    {op: [2.0 q0 * ]; expected: ( 2.0  4.0 neg  6.0      8.0    ) quaternion;}
+    {op: [q0 q1 *  ]; expected: (44.0 32.0     12.0 neg 46.0 neg) quaternion;}
+    {op: [q0 conj  ]; expected: ( 1.0  2.0 neg  3.0      4.0 neg) quaternion;}
+
+    {op: [q0 q1 dot]; expected: 18.0 neg;}
+
+    {op: [q0 unit                          ]; expected: q                           ;}
+    {op: [q0 unitChecked                   ]; expected: q                           ;}
+    {op: [q0 100.0 unitCheckedWithThreshold]; expected: (0.0 0.0 0.0 1.0) quaternion;}
+  );
+
+  tests [
+    t:;
+    t.op t.expected checkEqual
+  ] each
+] call
+
+# nlerp
+[
+  q0: (1.0 2.0 neg 3.0 4.0    ) quaternion;
+  q1: (5.0 6.0     7.0 8.0 neg) quaternion;
+  tests: (
+    {f: 0.0; q: q0 unit                                                                                        ;}
+    {f: 0.3; q: (0.129913960492326 neg 0.519655841969305 neg 0.0                   0.84444074320012) quaternion;}
+    {f: 0.6; q: (0.298083609185736 neg 0.504449184775861 neg 0.343942625983541 neg 0.73374426876488) quaternion;}
+    {f: 1.0; q: q1 unit                                                                                        ;}
+  );
+  2 [
+    tests [
+      t:;
+      q0 q1 t.f              nlerp t.q checkEqual
+      q1 q0 1 t.f cast t.f - nlerp t.q checkEqual
+    ] each
+    q1.entries neg @q1.!entries
+  ] times
+] call
+
+# slerp
+[
+  q0: (1.0 2.0 neg 3.0 4.0    ) quaternion unit;
+  q1: (5.0 6.0     7.0 8.0 neg) quaternion unit;
+  tests: (
+    {f: 0.0; q: q0 new                                                                                      ;}
+    {f: 0.3; q: (0.000501537327541 neg 0.48176120346404 neg 0.23987752707694     0.8428313373983) quaternion;}
+    {f: 0.6; q: (0.183499748964156 neg 0.52391886713953 neg 0.10504006435854 neg 0.8251081430121) quaternion;}
+    {f: 1.0; q: q1 new                                                                                      ;}
+  );
+  2 [
```

Comment:
Remove copypaste

Reply:
I can combine slerp and nlerp into one test function and extend tests structures like
```
tests: (
  {
   f:;
    q: {
     nlerp:;
     slerp:;
    }
  }
) 
```
Is it ok?

Reply:
I think it´s better to use helper function.